### PR TITLE
ci: add no-channel check

### PR DESCRIPTION
**Pull request by zimbatm**
###### Motivation for this change

Somebody merged a change in `nixos-20.09`. See #109384

###### Proposal

This is a proposal to prevent such issues in the future.

1. Add a CI check that always fails, that contains an explanation.
2. Configure nixpkgs to require this CI status to pass on the protected branches.

Here is how the error looks like:

image

And the branch protection change that would be put in place:
image

### CODE REVIEW

**Comment by Mic92** [comment]
I assume CI is needed because there is no different way to prevent people from pushing to these branches?

**Comment by SuperSandro2000** [comment]
> I assume CI is needed because there is no different way to prevent people from pushing to these branches?

I think admins ignore the branch proection.

**Comment by zowoq** [comment]
https://github.com/NixOS/ofborg/pull/546 does this with `ofborg` but I think this is better as we can "require" it which isn't possible with the `ofborg` check.

Being able to "require" `ofborg` checks has been mentioned previously but AFAIK there hasn't been anything since https://github.com/NixOS/nixpkgs/pull/87853#issuecomment-632677285.

File: .github/workflows/no-channel.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  fail:
+    name: "This PR is is targeting a channel branch"
+    runs-on: ubuntu-latest
+    steps:
+    - run: |
+        cat <<EOF
+        The nixos-* and nixpkgs-* branches are pushed to by the channel
+        release script and should not be merged into directly.
+
+        Please target the equivalent release-* branch instead.
```
Comment (cole-h) [inline .github/workflows/no-channel.yml:19]:
```suggestion
        Please target the equivalent release-* branch or master instead.
```

{nixos,nixpkgs}-unstable are built from master, so should probably call that out, too.

**Comment by Mic92** [comment]
> [NixOS/ofborg#546](https://github.com/NixOS/ofborg/pull/546) does similar to this with `ofborg` but I think this is better as we can "require" it which isn't possible with the `ofborg` check.
> 
> Being able to "require" `ofborg` checks has been mentioned previously but AFAIK there hasn't been anything since [#87853 (comment)](https://github.com/NixOS/nixpkgs/pull/87853#issuecomment-632677285).

I think the major concern there was that if ofborg would be not available the whole nixpkgs development would break. There are probably less concerns with github actions as there is a dedicated team with pager duty.

**Review by Mic92 — APPROVED** [review]

**Comment by zimbatm** [comment]
Tested and #130906 works as expected

**Pull request merged**